### ISSUE TITLE
Zoomed in software rendering does not properly invalidate balloons.

### ISSUE THREAD

**Issue by mrmbernardi** [OpenRCT2/OpenRCT2#23376]
### Operating System

Any

### OpenRCT2 build

OpenRCT2, v0.4.17-8

### Base game

RollerCoaster Tycoon 2

### Describe the issue

It seems that the invalidated region may not be scaling properly with the zoom level when zoomed in.

![Image](https://github.com/user-attachments/assets/2f86c7be-f1cb-43c5-b2d5-548dcacb511a)

### Steps to reproduce

Zoom in on software rendering and look for guests with balloons

**Comment by ZehMatt** [comment]
It's not just balloons, there is something really off about the invalidation.
![Image](https://github.com/user-attachments/assets/e130c3a0-4082-4e5a-a7b1-2c8e7bb8b59e)

The guests on the bridge aren't there anymore, that are left-over pixels mostly.

**Comment by mrmbernardi** [comment]
Is that without zoom?

**Comment by ZehMatt** [comment]
> Is that without zooming in? What zoom level is that?

Doesn't seem to matter.
![Image](https://github.com/user-attachments/assets/2d1bb2d4-d171-4ef2-8dcf-61e9e5549dbd)

**Comment by ZehMatt** [comment]
Regression from #23235

**Comment by AaronVanGeffen** [comment]
Thanks for investigating, @ZehMatt. I think I know what's wrong. 

Consider the following code at the bottom of the `Peep::Paint` function in Peep.cpp:
```cpp
    auto* guest = As<Guest>();
    if (guest != nullptr)
    {
        if (baseImageId >= kPeepSpriteHatStateWatchRideId && baseImageId < (kPeepSpriteHatStateSittingIdleId + 4))
        {
            imageId = ImageId(baseImageId + 32, guest->HatColour);
            PaintAddImageAsChild(session, imageId, offset, bb);
            return;
        }

        if (baseImageId >= kPeepSpriteBalloonStateWatchRideId && baseImageId < (kPeepSpriteBalloonStateSittingIdleId + 4))
        {
            imageId = ImageId(baseImageId + 32, guest->BalloonColour);
            PaintAddImageAsChild(session, imageId, offset, bb);
            return;
        }

        if (baseImageId >= kPeepSpriteUmbrellaStateWalkingId && baseImageId < (kPeepSpriteUmbrellaStateSittingIdleId + 4))
        {
            imageId = ImageId(baseImageId + 32, guest->UmbrellaColour);
            PaintAddImageAsChild(session, imageId, offset, bb);
            return;
        }
    }
```

Basically, if the peep is wearing a hat or holding a balloon or umbrella, we're painting another sprite on top. The inference algorithm introduced in #23235 currently has no way of knowing about these! Therefore, the negative height will always be off by 6-10 pixels for these three animation groups. We will need to account for this, somehow.

**Comment by mrmbernardi** [comment]
In the scene @ZehMatt posted, there's a section where the handyman and some other peeps near him are completely undrawn for a short section of time so I think something else is going on there too.

**Comment by ZehMatt** [comment]
> In the first scene [@ZehMatt](https://github.com/ZehMatt) posted, there's a section where the handyman and some other peeps near him are completely undrawn for a short section of time so I think something else is going on there too.

I investigated that further, I don't remember where I got this save from but it seems that the invalidation rect are imported from the save, after deleting all guests and staff and re-creating them the issue disappeared altho I'm not sure why it wouldn't update that data once it enters a new animation but I guess "walking" is just one set so it doesn't have to, we should adjust the data upon import and not trust saves to have it correct.

**Comment by AaronVanGeffen** [comment]
> we should adjust the data upon import and not trust saves to have it correct.

Agreed. I was under the impression that was already done tbh! Especially because headless clients don't have the graphics to infer the sprite bounds in the first place. Maybe that's how this issue arose in the first place.

**Comment by mrmbernardi** [comment]
IDK if it's related but there were some versions of the game where parks loaded using the S4 importer would have the peeps' invalidation rectangles set up incorrectly: https://github.com/OpenRCT2/OpenRCT2/pull/21125/files

Perhaps if you saved while they were incorrect, they'd remain incorrect. All the more reason why they should probably just be calculated at runtime.

**Comment by ZehMatt** [comment]
> > > In the first scene [@ZehMatt](https://github.com/ZehMatt) posted, there's a section where the handyman and some other peeps near him are completely undrawn for a short section of time so I think something else is going on there too.
> > 
> > 
> > I investigated that further, I don't remember where I got this save from but it seems that the invalidation rect are imported from the save, after deleting all guests and staff and re-creating them the issue disappeared altho I'm not sure why it wouldn't update that data once it enters a new animation but I guess "walking" is just one set so it doesn't have to, we should adjust the data upon import and not trust saves to have it correct.
> 
> if the save was made on a server, i.e. headless, that would happen.

Hm right, I think we should just stop exporting/importing this, its only really relevant for invalidation and clicking on peeps.

**Comment by AaronVanGeffen** [comment]
> Hm right, I think we should just stop exporting/importing this, its only really relevant for invalidation and clicking on peeps.

I think we should reassign them when loading a save file, indeed. I've already implemented this to an extent in #23328, mainly for SV4 import. This can easily be generalised, though. Let's go with that. I'll file a PR later this week.

**Comment by AaronVanGeffen** [comment]
Filed a fix in #23425
